text/0000-macro-shorthand.md

```diff
@@ -0,0 +1,101 @@
+- Feature Name: `macro_shorthand`
+- Start Date: 2022-05-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This is a proposal to make `m!123` a short-hand for `m!(123)`.
+
+```rust
+    let num = bignum!12345678901234567890123456789012345678901234567890;
+    let msg = f!"n = {num}";
+    let file = w!"C:\\Thing";
+```
+
+# Motivation
+[motivation]: #motivation
+
+In the Rust 2021 edition we reserved all prefixes for literals, so we can give them a meaning in the future.
+However, many ideas for specific literal prefixes (e.g. for wide strings or bignums) are domain- or crate-specific,
+and should arguably not be a builtin part of the language itself.
+
+By making `m!123` a short-hand for `m!(123)`, we get a syntax that's just as convenient and light-weight as built-in prefixes,
+but through a mechanism that allows them to be user-defined, without any extra language features necessary to define them.
+
+For example:
+
+- Those who want "f-strings" can then simply do `use std::format as f;` and then use `f!"{a} {b}"`.
+- Windows crates could provide wide strings using `w!"C:\\"`
+- An arbitrary precision number crate could provide `bignum!12345678901234567890123456789012345678901234567890`.
+
+The difference with `f!("{a} {b}")`, `w!("C:\\")` and `bignum!(123...890)` is small, but significant.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Macros can be invoked using `m!(..)`, `m![..]`, `m!{..}` or `m!..` syntax.
+In the last case, the argument must be a single literal, such as `m!123`, `m!2.1`, `m!"abc"`, or `m!'x'`.
+From the perspective of a macro definition, these are all identical, and a macro cannot differentiate between the different call syntaxes.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The macro invocation syntax is changed from
+
+```
+MacroInvocation :
+   SimplePath ! DelimTokenTree
+
+MacroInvocationSemi :
+     SimplePath ! ( TokenTree* ) ;
+   | SimplePath ! [ TokenTree* ] ;
+   | SimplePath ! { TokenTree* }
+```
+
+to
+
+```
+MacroInvocation :
+     SimplePath ! Literal
```

Comment:
Is this `Literal` unambiguous enough to distinguish between these two situations?
* "literal" *token*, aka `proc_macro::Literal` (string/character/numeric literals)
* "literal" (e.g. expression) *grammar*, aka `rustc_ast::Literal` aka `$lit:literal` macro inputs
  * besides what literal tokens support, this also includes `false` and `true`
  * also it does more validation (suffixes, presumably string escapes, forcing integers into `u128`, etc.)

I would assume the former (esp. given the mention of `m!identifier` later in the RFC, where arguably `m!false`/`m!true` would fit), but `$lit:literal` being the latter muddles the waters a bit sadly.

This is confusing enough that the reference is wrongly mentioning `false`/`true` on stable under "tokens" (was fixed since by https://github.com/rust-lang/reference/pull/1189).

(Thanks to @solson for bringing up the potential ambiguity wrt `bool` literals - I would've naively assumed it was a non-concern at first)

Reply:
The first one.

`$lit:literal` doesn't accept very large integers, which is important for use cases like `bignum!123`.

Allowing `json!true` and `json!false` might be reasonable, but then we also need to have a discussion about `json!null`, `js!NaN`, `py!True`, and so on, which would all need the `identifier` or `path` grammer. So I'd like to leave that discussion for a future RFC.

Reply:
> `$lit:literal` doesn't accept very large integers, which is important for use cases like `bignum!123`

I wonder if we could fix that?

Not a blocker for this RFC, but we *could* allow arbitrary-length integer literals to get fed into macros, and only check if they fit when the resulting token stream *from* the macro gets parsed.

Reply:
`$x:tt` accepts arbitrary length integers just fine. (See the playground link above.) It's just that non-`tt` things like `$x:literal` no longer represent the raw tokens but instead a portion of already processed/parsed AST.